From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Andrii Sultanov <andriy.sultanov@vates.tech>
Date: Thu, 17 Sep 2026 08:12:14 +0000
Subject: [PATCH] vhd-tool: Pipeline the writes in stream_nbd

Instead of immediately blocking on every write, schedule several writes at once
and wait on them with Lwt.nchoose.

The NBD library can already handle several writes at the same time - we only
need to keep track of allocated buffers and re-use them when promises complete.

(Lwt does not pre-empt promises, so no additional synchronization is necessary)

Bring the twomib constant out to the module level, remove its re-definitions
from inside functions.

Tracking-reference: XCPNG-3820
Signed-off-by: Andrii Sultanov <andriy.sultanov@vates.tech>
---
 ocaml/vhd-tool/src/impl.ml | 74 +++++++++++++++++++++++++++++++-------
 quality-gate.sh            |  2 +-
 2 files changed, 63 insertions(+), 13 deletions(-)

diff --git a/ocaml/vhd-tool/src/impl.ml b/ocaml/vhd-tool/src/impl.ml
index 5b74ec474..a8127944f 100644
--- a/ocaml/vhd-tool/src/impl.ml
+++ b/ocaml/vhd-tool/src/impl.ml
@@ -44,6 +44,8 @@ end)
 
 open F
 
+let twomib = 2 * 1024 * 1024
+
 let require name arg =
   match arg with
   | None ->
@@ -245,18 +247,67 @@ let stream_nbd _common c s prezeroed ~export ?(progress = no_progress_bar) () =
   in
   let p = progress total_work in
 
+  (* Pipeline several requests at the same time.
+     The NBD client already handles several independent requests in parallel.
+     Allocate [depth] buffers and re-use them once each promise completes. *)
+  let depth = 8 in
+  let pool = Array.init depth (fun _ -> IO.alloc twomib) in
+  let free = ref (List.init depth Fun.id) in
+  (* A list of promises and integers wrapped in an option representing an
+     index into the pool to allow re-using the buffers.
+     expand_empty allocates a single zero-filled buffer and does not need to
+     re-use the ones from the pool, so index is [None] in that case *)
+  let inflight : (unit Lwt.t * int option) list ref = ref [] in
+  let reap () =
+    match !inflight with
+    | [] ->
+        return ()
+    | l ->
+        Lwt.nchoose_split (List.map fst l) >>= fun (_, pending) ->
+        let pending, completed =
+          List.partition_map
+            (fun req ->
+              let t, i = req in
+              if List.memq t pending then
+                Either.Left req
+              else
+                Either.Right i
+            )
+            l
+        in
+        inflight := pending ;
+        let reusable_buffers = List.filter_map Fun.id completed in
+        free := reusable_buffers @ !free ;
+        return ()
+  in
+  let take_free_buffer () =
+    (* [reap] blocks until at least one promise succeeds or fails,
+       [free] list is guaranteed to be non-empty after calling it *)
+    (if !free = [] then reap () else return ()) >>= fun () ->
+    let index = List.hd !free in
+    free := List.tl !free ;
+    let buf = pool.(index) in
+    return (buf, Some index)
+  in
+  let rec drain () =
+    if !inflight = [] then return () else reap () >>= fun () -> drain ()
+  in
+
   (if not prezeroed then expand_empty s else return s) >>= fun s ->
-  expand_copy s >>= fun s ->
+  expand_copy s ~get_buffer:take_free_buffer >>= fun s ->
   fold_left
     (fun (sector, work_done) x ->
       ( match x with
-        | `Sectors (data, _) -> (
-            Client.write server (Int64.mul sector 512L) [data] >>= function
-            | Ok () ->
-                return Int64.(of_int (Cstruct.length data))
-            | Error _e ->
-                fail (Failure "Got error from NBD library")
-          )
+        | `Sectors (data, i) ->
+            let t =
+              Client.write server (Int64.mul sector 512L) [data] >>= function
+              | Ok () ->
+                  return ()
+              | Error _e ->
+                  fail (Failure "Got error from NBD library")
+            in
+            inflight := (t, i) :: !inflight ;
+            return Int64.(of_int (Cstruct.length data))
         | `Empty _n ->
             (* must be prezeroed *)
             assert prezeroed ;
@@ -270,6 +321,8 @@ let stream_nbd _common c s prezeroed ~export ?(progress = no_progress_bar) () =
               )
         )
       >>= fun work ->
+      (* Progress is reported when buffers are scheduled to be written,
+         not when completed *)
       let sector = Int64.add sector (Vhd_format.Element.len x) in
       let work_done = Int64.add work_done work in
       p work_done ;
@@ -277,6 +330,7 @@ let stream_nbd _common c s prezeroed ~export ?(progress = no_progress_bar) () =
     )
     (0L, 0L) s.elements
   >>= fun _ ->
+  drain () >>= fun () ->
   p total_work ;
 
   return (Some total_work)
@@ -617,7 +671,6 @@ let serve_vhd_to_raw total_size c dest prezeroed progress _ _ =
 
 let serve_tar_to_raw total_size c dest prezeroed progress expected_prefix
     ignore_checksums =
-  let twomib = 2 * 1024 * 1024 in
   let buffer = IO.alloc twomib in
   let header = IO.alloc 512 in
 
@@ -1224,7 +1277,6 @@ let serve_nbd_to_raw common size c dest _ _ _ _ =
   let buf = Cstruct.create (Negotiate.sizeof `V1) in
   Negotiate.marshal buf (Negotiate.V1 {Negotiate.size; flags}) ;
   c.Channels.really_write buf >>= fun () ->
-  let twomib = 2 * 1024 * 1024 in
   let block = IO.alloc twomib in
   let inblocks fn request =
     let rec loop offset remaining =
@@ -1281,7 +1333,6 @@ let serve_nbd_to_raw common size c dest _ _ _ _ =
 
 let serve_chunked_to_raw _ c dest _ _ _ _ =
   let header = Cstruct.create Chunked.sizeof in
-  let twomib = 2 * 1024 * 1024 in
   let buffer = IO.alloc twomib in
   let rec loop () =
     c.Channels.really_read header >>= fun () ->
@@ -1316,7 +1367,6 @@ let round_up_to_sector unbuffered len =
     len
 
 let serve_raw_to_raw common size c dest _ progress _ _ =
-  let twomib = 2 * 1024 * 1024 in
   let buffer = IO.alloc twomib in
   let p = progress size in
   let rec loop offset remaining =
diff --git a/quality-gate.sh b/quality-gate.sh
index 7143abe02..0ff70a541 100755
--- a/quality-gate.sh
+++ b/quality-gate.sh
@@ -3,7 +3,7 @@
 set -e
 
 list-hd () {
-  N=239
+  N=240
   LIST_HD=$(git grep -r --count 'List.hd' -- **/*.ml | cut -d ':' -f 2 | paste -sd+ - | bc)
   if [ "$LIST_HD" -eq "$N" ]; then
     echo "OK counted $LIST_HD List.hd usages"
